Added SecBI integration

RELnemtzov opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
Ready/In Progress/In Hold(Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------


## Required version of Demisto
x.x.x

## Does it break backward compatibility?
   - Yes
       - Further details:
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [ ] Dependency 1
- [ ] Dependency 2
- [ ] Dependency 3

## Additional changes
Describe additional changes done, for example adding a function to common server.

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](link)
- [ ] [CHANGELOG](link)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=4784) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=4784) before we can accept your contribution.<br/><hr/>**Ariel Nemtzov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=4784) it.</sub>

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+            client.test_module(API_URL, API_KEY)
+            demisto.results('ok')
+        elif demisto.command() == 'secbi-get-incidents-list':
+            query = args.get('query', None)
```
Please wrap the content of the condition in a command function which returns the outputs, and call it from here with `return_outputs`
Look [here](https://github.com/demisto/content/blob/master/Templates/Integrations/Authentication/Authentication.py#L447) for an example

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+                'SecBI.IncidentsList': incidents_list
+            }
+            return_outputs(human_readable, entry_context, incidents_list)
+        elif demisto.command() == 'secbi-get-incident':
```
Please wrap the content of the condition in a command function which returns the outputs, and call it from here with `return_outputs`
Look [here](https://github.com/demisto/content/blob/master/Templates/Integrations/Authentication/Authentication.py#L447) for an example

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+                'SecBI.Incident(val.id == obj.id)': incident_data
+            }
+            return_outputs(human_readable, entry_context, incident_data)
+        elif demisto.command() == 'secbi-get-incident-by-host':
```
Please wrap the content of the condition in a command function which returns the outputs, and call it from here with `return_outputs`
Look [here](https://github.com/demisto/content/blob/master/Templates/Integrations/Authentication/Authentication.py#L447) for an example

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+            LOG("Error when fetching Incidents list from SecBI: {}".format(e.args[0]))
+            LOG.print_log()
+            raise
+        return json.loads(ids_list_str)
```
Why not just send `resp_type=json` if we're doing here `json.loads()` here anyway ?

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+                                              url_suffix=SECBI_API_ENDPOINT_INCIDENTS,
+                                              params=params,
+                                              resp_type='text')
+        except Exception as e:
```
`_http_request` already handles exceptions, no need for that.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+                                              params=params)
+            secbi_incident = SecBIIncident(raw_incident)
+            return vars(secbi_incident)
+        except Exception as e:
```
`_http_request` already handles exceptions, no need for that.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+            raw_incident = self._http_request(method='GET', url_suffix=SECBI_API_ENDPOINT_INCIDENT, params=params)
+            secbi_incident = SecBIIncident(raw_incident)
+            return vars(secbi_incident)
+        except Exception as e:
```
`_http_request` already handles exceptions, no need for that.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+        """
+        try:
+            self._http_request(method='GET', url_suffix=SECBI_API_ENDPOINT_STATUS)
+        except Exception as e:
```
`_http_request` already handles exceptions, no need for that.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+                                              url_suffix='{}/{}'.format(SECBI_API_ENDPOINT_INCIDENT, incident_id),
+                                              params=params)
+            secbi_incident = SecBIIncident(raw_incident)
+            return vars(secbi_incident)
```
I think this is a bit of an overkill here, why not init a dictionary and map the incident fields to it?

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+        try:
+            raw_incident = self._http_request(method='GET', url_suffix=SECBI_API_ENDPOINT_INCIDENT, params=params)
+            secbi_incident = SecBIIncident(raw_incident)
+            return vars(secbi_incident)
```
I think this is a bit of an overkill here, why not init a dictionary and map the incident fields to it?

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,197 @@
+''' COMMANDS MANAGER / SWITCH PANEL '''
+
+
+def get_human_readable_incidents_list(incidents_list):
```
Don't really see the added value of a function of 2 lines which is called only once, let's just insert those 2 lines to where this function is called.

RELnemtzov commented:
> 1. Please add [test](https://github.com/demisto/content/tree/master/docs/tests)
Does that mean that I'll need to have a live SecBI instance in the cloud at all times to answer the integration tests run by your CircleCI, presumably every night?

Itay4 commented:
> > 1. Please add [test](https://github.com/demisto/content/tree/master/docs/tests)
> 
> Does that mean that I'll need to have a live SecBI instance in the cloud at all times to answer the integration tests run by your CircleCI, presumably every night?

correct

RELnemtzov commented:
Completed code fixes. Still need to add the tests and documentation.
Not entirely clear on adding the documentation. As per [this](https://github.com/demisto/content/tree/master/docs/integration_documentation#posting-documentation) link that you posted, I need to place the generated documentation in the "integration/automation script package directory":

- Where is that directory?
- What needs to placed there? The documentation HTML file I downloaded from the generation?

Itay4 commented:
> Completed code fixes. Still need to add the tests and documentation.
> Not entirely clear on adding the documentation. As per [this](https://github.com/demisto/content/tree/master/docs/integration_documentation#posting-documentation) link that you posted, I need to place the generated documentation in the "integration/automation script package directory":
> 
> * Where is that directory?
> * What needs to placed there? The documentation HTML file I downloaded from the generation?

The documentation should be placed in the integration directory, which is `Integrations/SecBI` in this case in file named `README.md`, which its content is the HTML file.

Itay4 commented:
please pull from `contrib/RELnemtzov_master `

RELnemtzov commented:
Added test and documentation. Updated the PR from `content:master`.
I see some checks were not successful. Please advise as to what I'm missing.
Thanks.

Itay4 commented:
@RELnemtzov please see my comment above

RELnemtzov commented:
> @RELnemtzov please see my comment above

Regarding pulling from `contrib/RELnemtzov_master`? I did that, and merged it in to my branch. It says I'm already up to date.

Itay4 commented:
> > @RELnemtzov please see my comment above
> 
> Regarding pulling from `contrib/RELnemtzov_master`? I did that, and merged it in to my branch. It says I'm already up to date.

@RELnemtzov are you sure? there still is a big diff and a lot of changed files

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+''' HELPER FUNCTIONS '''
+
+
+def capitalize(s: str):
```
Use Python [builtin capitalize function](https://docs.python.org/3/library/stdtypes.html#str.capitalize) instead

RELnemtzov replied:
The built-in capitalize will lowercase the rest of the string. I only want to capitalize the first character and leave the rest as is. For example, I want `firstAppearance` to become `FirstAppearance` and not `Firstappearance`

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+    # Log exceptions
+    except Exception as e:
+        msg = getattr(e, 'message', repr(e))
+        LOG(msg)
```
No need for this, already done in `return_error`

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+    except Exception as e:
+        msg = getattr(e, 'message', repr(e))
+        LOG(msg)
+        return_error(msg, error=e)
```
```suggestion
        return_error(str(e), error=e)
```

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+    return s[0].upper() + s[1:]
+
+
+def get_dict_without_none(**kwargs):
```
Please use `assign_params` function instead

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+        raw_incident = self._http_request(method='GET',
+                                          url_suffix='{}/{}'.format(SECBI_API_ENDPOINT_INCIDENT, incident_id),
+                                          params=params)
+        return as_secbi_incident(raw_incident)
```
Please call `as_secbi_incident` from the command function

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+        """
+        params = get_dict_without_none(host=host, fields=SECBI_INCIDENT_FIELDS)
+        raw_incident = self._http_request(method='GET', url_suffix=SECBI_API_ENDPOINT_INCIDENT, params=params)
+        return as_secbi_incident(raw_incident)
```
Please call `as_secbi_incident` from the command function

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+
+    human_readable = tableToMarkdown('SecBI incident ID "{}"'.format(incident_id), incident_data)
+    entry_context = {
+        'SecBI.Incident(val.id == obj.id)': incident_data
```
```suggestion
        'SecBI.Incident(val.ID === obj.ID)': incident_data
```

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+
+    human_readable = tableToMarkdown('SecBI incident by host "{}"'.format(host), incident_data)
+    entry_context = {
+        # ID comparison might be problematic if the host changes Incidents and the ID changes
```
Not sure I understand that comment, can you please elaborate?

RELnemtzov replied:
This has to do with the way hosts are handled within SecBI. A given `host` may move change incidents between any two given SecBI "runs" (or digestions of data), thus the ID (and content) will change although they have the same desired host.
This might cause some problems if a user is relying on the same ID for the same host.

Itay4 replied:
In that case, we will have to change the unique field from `ID` to one that does not change

RELnemtzov replied:
This is by design according to the use-case defined. The comment was more for us to take note in case the use-case definition ever changes.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,203 @@
+    human_readable = tableToMarkdown('SecBI incident by host "{}"'.format(host), incident_data)
+    entry_context = {
+        # ID comparison might be problematic if the host changes Incidents and the ID changes
+        'SecBI.Incident(val.id == obj.id)': incident_data
```
```suggestion
        'SecBI.Incident(val.ID === obj.ID)': incident_data
```

Itay4 commented on Integrations/SecBI/SecBI_description.md:
```diff
@@ -0,0 +1,6 @@
+The SecBI integration for Demisto requires the configuration of the following parameters:
+
+* SECBI_API_URL - The URL to your SecBI instance which will be used to query SecBI, such as: https://<SUBDOMAIN>.secbi.com
+* SECBI_API_KEY - The API key used for authentication with your SecBI instance
```
Can you elaborate on how to retrieve the API key ?

Itay4 commented:
@kirbles19 please review the following:
- [ ] https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/SecBI.yml
- [ ] https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/SecBI_description.md

see the [README](https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/README.md) for reference

RELnemtzov commented on Tests/conf.json:
```diff
@@ -1940,6 +1940,14 @@
+        {
+            "integrations": "SecBI",
+            "playbookID": "SecBI-Test",
+            "params": {
```
Is that what this means from [here]( https://github.com/demisto/content/tree/master/docs/tests#adding-tests-to-confjson):
If your integration must be configured to be executed, you must add the code below to the content-test-conf/conf.json file. The field names must match the parameters that you assigned to your integration.
{
  "name": "carbonblack-v2",
  "params": {
    "serverurl": "https://example.com:30035/",
    "apitoken": "exampleapikey",
    "insecure": true,
    "proxy": false
  }
}

RELnemtzov commented:
I've already signed the CLA. When it asks me to Sign in to GitHub to Sign, I do and it won't let me sign it again and says "You have signed the CLA for multiple repositories or organizations".

Might be something similar to [this](https://github.com/cla-assistant/cla-assistant/issues/511) issue with the cla-assistant?

Itay4 commented:
> I've already signed the CLA. When it asks me to Sign in to GitHub to Sign, I do and it won't let me sign it again and says "You have signed the CLA for multiple repositories or organizations".
> 
> Might be something similar to [this](https://github.com/cla-assistant/cla-assistant/issues/511) issue with the cla-assistant?

What is the user `Ariel Nemtzov`? you should sign with `RELnemtzov`

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,193 @@
+    formatted = [dict(ID=s) for s in incidents_list]
+    human_readable = tableToMarkdown('List of SecBI Incidents', formatted)
+    entry_context = {
+        'SecBI.IncidentsList': incidents_list
```
```suggestion
        'SecBI.Incident': incidents_list
```

RELnemtzov replied:
This isn't the same thing. The `IncidentsList` only contains a list of IDs (strings). Where as the `Incident` contains a JSON of an actual Incident's data.

Itay4 replied:
arent these the same objects? like same incidents? the ids are the same ?

RELnemtzov replied:
Each ID in the `IncidentsList` corresponds to an actual SecBI Incident that can be requested via the `secbi-get-incident` command and then the actual Incident JSON will be retrieved and saved to the `Incident` context.

Itay4 replied:
then both commands should populate outputs under the same context path: `SecBI.Incident`

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,193 @@
+    incident_data = as_secbi_incident(client.secbi_get_incident(incident_id))
+
+    human_readable = tableToMarkdown('SecBI incident ID "{}"'.format(incident_id), incident_data)
+    entry_context = {
```
Please also populate output according to [IP context standard](https://github.com/demisto/content/tree/master/docs/context_standards#ip) and [Host context standard](url)

RELnemtzov replied:
These were not currently planned use-cases by our product team at this time.

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,193 @@
+    incident_data = as_secbi_incident(client.secbi_get_incident_by_host(host))
+
+    human_readable = tableToMarkdown('SecBI incident by host "{}"'.format(host), incident_data)
+    entry_context = {
```
Please also populate output according to [IP context standard](https://github.com/demisto/content/tree/master/docs/context_standards#ip) and [Host context standard](url)

RELnemtzov replied:
These were not currently planned use-cases by our product team at this time.

RELnemtzov commented:
> Got two unexpected results:
Fixed this issue.

Still the same problem with the CLA.

glicht commented:
@RELnemtzov for the CLA to pass please add to your github account the email address you first used to do the initial commits: `ariel.nemtzov at secbi.com`. You can add this by following the instructions here: https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account

richardbluestone commented:
> @kirbles19 please review the following:
> 
> * [ ]  https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/SecBI.yml
> * [ ]  https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/SecBI_description.md
> 
> see the [README](https://github.com/demisto/content/blob/ab3eda17695d5213ed8dc23bbd2b7baa5ebfdc70/Integrations/SecBI/README.md) for reference

I've reviewed the yml and md files.

Itay4 commented:
getting the following non-indicative error messages:
![image](https://user-images.githubusercontent.com/31018228/69156781-a2510600-0aec-11ea-98a4-089ad6a227ed.png)

![image](https://user-images.githubusercontent.com/31018228/69156814-b4cb3f80-0aec-11ea-84d4-e4ff1bb11bfc.png)

RELnemtzov commented:
> getting the following non-indicative error messages:
> ![image](https://user-images.githubusercontent.com/31018228/69156781-a2510600-0aec-11ea-98a4-089ad6a227ed.png)
> 
> ![image](https://user-images.githubusercontent.com/31018228/69156814-b4cb3f80-0aec-11ea-84d4-e4ff1bb11bfc.png)

Clarified the error messages and squashed the previous fixes' commits.

Itay4 commented:
![image](https://user-images.githubusercontent.com/31018228/69237436-5f4f6b00-0b9e-11ea-99e1-c3f8e5e2f0e5.png)

Limit doesn't seem to work as expected

Itay4 commented on Integrations/SecBI/SecBI.yml:
```diff
@@ -0,0 +1,122 @@
+  commands:
+  - arguments:
+    - default: false
+      description: The Query for which to filter the incident IDs.
```
please mention here its elastic, and add an example

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,196 @@
+    entry_context = {
+        'SecBI.Incident(val.ID === obj.ID)': incident_data
+    }
+    return human_readable, entry_context, incident_data
```
the third object returned here (`incident_data`) should be the raw response returned from the API

Itay4 commented on Integrations/SecBI/SecBI.py:
```diff
@@ -0,0 +1,196 @@
+        # ID comparison might be problematic if the host changes Incidents and the ID changes
+        'SecBI.Incident(val.ID === obj.ID)': incident_data
+    }
+    return human_readable, entry_context, incident_data
```
the third object returned here (`incident_data`) should be the raw response returned from the API

Itay4 commented:
![image](https://user-images.githubusercontent.com/31018228/69238467-c66e1f00-0ba0-11ea-827f-0809f5dd1aaf.png)
doesn't seem to be the expected result, or am I using it wrong?

RELnemtzov commented:
> ![image](https://user-images.githubusercontent.com/31018228/69238467-c66e1f00-0ba0-11ea-827f-0809f5dd1aaf.png)
> doesn't seem to be the expected result, or am I using it wrong?

You were using it correctly I solved an issue with the API server that was answering incorrectly.

Itay4 commented:
Please improve error messages:
![image](https://user-images.githubusercontent.com/31018228/69634146-a49cee00-101f-11ea-98ec-6ca91c928c2b.png)

![image](https://user-images.githubusercontent.com/31018228/69634298-ec237a00-101f-11ea-9533-597e3ce919b7.png)

![image](https://user-images.githubusercontent.com/31018228/69634317-f34a8800-101f-11ea-8719-0cc00ef290cc.png)

RELnemtzov commented:
Sorry for the delay, I'll need a couple more days before I can get back to this.

anara123 commented:
@RELnemtzov sure, let us know when you have time to continue.

RELnemtzov commented:
> Please improve error messages:
These were solved by fixing a server-side bug, which wasn't generating error responses properly.

I think that should conclude all remarks, right?
